src/serializer/Referentializer.js

```diff
@@ -194,8 +194,11 @@ export class Referentializer {
 
           // Already referentialized in prior scope
           if (residualBinding.declarativeEnvironmentRecord) {
-            invariant(residualBinding.scope);
-            instance.scopeInstances.add(residualBinding.scope);
+            const { scope } = residualBinding;
+            invariant(scope);
+            if ([...instance.scopeInstances.values()].filter(s => s.name === scope.name).length === 0) {
```

Comment:
This is overly clever :), simply looping through with something like 

```javascript
for (const scope of instance.scopeInstances) {
 //....
}
```
is more clear.